File: Problem Set Volumes/Volume 7/706 - LC-Display/p706.java
/*
* Online Judge: Problem 706 - LC-Display
*
* Judge Version: JAVA 1.8.0 - OpenJDK Java
* Best Run Time: 0.180 Seconds
*
*/


import java.io.*;
import java.util.*;

/* Seven-segment representation for digits 0-9.
 Order of segments: top, upper-left, upper-right,
 middle, lower-left, lower-right, bottom */
public class Main {
    static final int[][] DIGIT_MAP = {
        {1,1,1,0,1,1,1}, // 0
        {0,0,1,0,0,1,0}, // 1
        {1,0,1,1,1,0,1}, // 2
        {1,0,1,1,0,1,1}, // 3
        {0,1,1,1,0,1,0}, // 4
        {1,1,0,1,0,1,1}, // 5
        {1,1,0,1,1,1,1}, // 6
        {1,0,1,0,0,1,0}, // 7
        {1,1,1,1,1,1,1}, // 8
        {1,1,1,1,0,1,1} // 9
    };

    public static void main(String[] args) throws Exception {

        // For local testing with a text file including sample input
        // Uncomment to use local testing
        // File testFile = new File("p706.txt");
        //if (testFile.exists()) {
        //    System.setIn(new FileInputStream(testFile));
        //}

        FastScanner fs = new FastScanner(System.in);
        StringBuilder out = new StringBuilder();


        while (true) {
            Integer sObj = fs.nextInt();
            //if (sObj == null) break;

            // Display size
            int s = sObj;

            // Number to render
            String n = fs.next();
            
            // End of input
            if (s == 0 && "0".equals(n)) break;

            // Total number of rows in a digit of size s
            int rows = 2 * s + 3;

            // Generate the display one row at a time
            for (int r = 0; r < rows; r++) {

                // Render each digit in the input number
                for (int num = 0; num < n.length(); num++) {

                    // Retrieve the seven-segment pattern for this digit
                    int d = n.charAt(num) - '0';

                    // Top horizontal segment
                    if (r == 0) {
                        out.append(' ');
                        for (int k = 0; k < s; k++) out.append(DIGIT_MAP[d][0] == 1 ? '-' : ' ');
                        out.append(' ');
                    // Upper vertical segments
                    } else if (r <= s) {
                        out.append(DIGIT_MAP[d][1] == 1 ? '|' : ' ');
                        for (int k = 0; k < s; k++) out.append(' ');
                        out.append(DIGIT_MAP[d][2] == 1 ? '|' : ' ');
                    // Middle horizontal segment
                    } else if (r == s + 1) {
                        out.append(' ');
                        for (int k = 0; k < s; k++) out.append(DIGIT_MAP[d][3] == 1 ? '-' : ' ');
                        out.append(' ');
                    // Lower vertical segments 
                    } else if (r < rows - 1) {
                        out.append(DIGIT_MAP[d][4] == 1 ? '|' : ' ');
                        for (int k = 0; k < s; k++) out.append(' ');
                        out.append(DIGIT_MAP[d][5] == 1 ? '|' : ' ');
                    // Bottom horizontal segment
                    } else {
                        out.append(' ');
                        for (int k = 0; k < s; k++) out.append(DIGIT_MAP[d][6] == 1 ? '-' : ' ');
                        out.append(' ');
                    }
                    // Space between digits
                    if (num + 1 < n.length()) out.append(' ');
                }
                out.append('\n');
            }
            out.append('\n');
        }

        System.out.print(out.toString());
    }

    // Implement fast scanner for input
    static class FastScanner {
        private final InputStream in;
        private final byte[] buffer = new byte[1 << 16];
        private int ptr = 0, len = 0;

        FastScanner(InputStream is) {
            this.in = is;
        }

        private int read() throws IOException {
            if (ptr >= len) {
                len = in.read(buffer);
                ptr = 0;
                if (len <= 0) return -1;
            }
            return buffer[ptr++];
        }

        Integer nextInt() throws IOException {
            int c;
            do {
                c = read();
            } while (c <= ' ' && c != -1);
            if (c == -1) return null;

            int sign = 1;
            if (c == '-') {
                sign = -1;
                c = read();
            }

            int val = 0;
            while (c > ' ') {
                val = val * 10 + (c - '0');
                c = read();
            }
            return val * sign;
        }

        String next() throws IOException {
            int c;
            do {
                c = read();
            } while (c <= ' ' && c != -1);
            if (c == -1) return null;

            StringBuilder sb = new StringBuilder();
            while (c > ' ') {
                sb.append((char)c);
                c = read();
            }
            return sb.toString();
        }
    }
}

File: Problem Set Volumes/Volume 7/706 - LC-Display/p706.py
"""
Online Judge: Problem 706 - LC-Display

Judge Version: PYTH3 3.5.1 - Python 3
Best Run Time: 0.040 Seconds

"""

import sys

# For local testing with a text file including sample input
# Uncomment to use local testing
#if sys.stdin.isatty():
#    sys.stdin = open("pc164_inputs.txt", "r")


# Seven-segment representation for digits 0-9.
# Order of segments: top, upper-left, upper-right,
# middle, lower-left, lower-right, bottom
DIGIT_MAP = [
    (1,1,1,0,1,1,1),  # 0
    (0,0,1,0,0,1,0),  # 1
    (1,0,1,1,1,0,1),  # 2
    (1,0,1,1,0,1,1),  # 3
    (0,1,1,1,0,1,0),  # 4
    (1,1,0,1,0,1,1),  # 5
    (1,1,0,1,1,1,1),  # 6
    (1,0,1,0,0,1,0),  # 7
    (1,1,1,1,1,1,1),  # 8
    (1,1,1,1,0,1,1),  # 9
]


def render(s, n):

    # Total number of rows in a digit of size s
    rows = 2 * s + 3

    # Stores every row of the final LCD display
    out = []

    # Generate the display one row at a time
    for r in range(rows):
 
        # Holds the current row for all digits
        row = []

        # Render each digit in the input number
        for num in n:
            
            # Retrieve the seven-segment pattern for this digit
            a, b, c, d, e, f, g = DIGIT_MAP[ord(num) - 48]

            # Top horizontal segment
            if r == 0:
                row.append(' ' + ('-' * s if a else ' ' * s) + ' ')

            # Upper vertical segments
            elif 1 <= r <= s:
                row.append(('|' if b else ' ') + (' ' * s) + ('|' if c else ' '))

            # Middle horizontal segment
            elif r == s + 1:
                row.append(' ' + ('-' * s if d else ' ' * s) + ' ')

            # Lower vertical segments 
            elif s + 2 <= r <= 2 * s + 1:
                row.append(('|' if e else ' ') + (' ' * s) + ('|' if f else ' '))

            # Bottom horizontal segment
            else:
                row.append(' ' + ('-' * s if g else ' ' * s) + ' ')

        # Combine all digits for this row
        out.append(' '.join(row))

    # Return the complete LCD display
    return '\n'.join(out)

# Read the entire input at once for faster processing
data = sys.stdin.buffer.read().split()
i = 0

#first = True

# Stores the rendered output for every input row
final_output = []


# Process test cases until the terminating pair "0 0" is encountered
while True:

    # Display size
    s = int(data[i])

    # Number to render
    n = data[i + 1].decode()
 
    i += 2

    # End of input
    if s == 0 and n == '0':
        break

    # Render the current input row
    final_output.append(render(s, n))

# Print the entire output
# Print a blank line between consecutive test cases
sys.stdout.write('\n\n'.join(final_output) + '\n\n')